File: JICS_MPI_11/C/group/group.c
#include "mpi.h"
#include <stdio.h>
#define NPROCS 12

int main(argc,argv)
int argc;
char *argv[];

{

  int rank, new_rank, sendbuf, recvbuf, numtasks;
  int ranks1[6]={0,1,2,3,4,5}, ranks2[6]={6,7,8,9,10,11};
  MPI_Group  orig_group, new_group;
  MPI_Comm   new_comm;

   /*----------------*/
  /* Initialize MPI */
 /*----------------*/

  MPI_Init(&argc,&argv);

   /*------------------------------------------------------*/
  /* Get my rank in the MPI_COMM_WORLD communicator group */
 /*------------------------------------------------------*/

  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   /*-------------------------------------------------------*/                                                                     
  /* Get the size of the MPI_COMM_WORLD communicator group */
 /*-------------------------------------------------------*/

  MPI_Comm_size(MPI_COMM_WORLD, &numtasks);

  if (numtasks != NPROCS) {
    printf("Must specify MP_PROCS= %d. Terminating.\n",NPROCS);
    MPI_Finalize();
    exit(0);
  }

  sendbuf = rank;

   /*-----------------------------------*/                                                                     
  /* Extract the original group handle */
 /*-----------------------------------*/

  MPI_Comm_group(MPI_COMM_WORLD, &orig_group);

   /*-------------------------------------------------------*/                                                                     
  /* Divide tasks into two distinct groups based upon rank */
 /*-------------------------------------------------------*/

  if (rank < NPROCS/2)
    MPI_Group_incl(orig_group, NPROCS/2, ranks1, &new_group);
  else
    MPI_Group_incl(orig_group, NPROCS/2, ranks2, &new_group);

   /*------------------------------------------------------------------------*/                                                                     
  /* Create new new communicator and then perform collective communications */
 /*------------------------------------------------------------------------*/

  MPI_Comm_create(MPI_COMM_WORLD, new_group, &new_comm);
  MPI_Allreduce(&sendbuf, &recvbuf, 1, MPI_INT, MPI_SUM, new_comm);


   /*-----------------------------------------------------*/
  /* Get my rank in the new new_group communicator group */
 /*-----------------------------------------------------*/

  MPI_Group_rank (new_group, &new_rank);

  printf("rank= %d newrank= %d recvbuf= %d\n",rank,new_rank,recvbuf);

   /*--------------*/
  /* Finalize MPI */
 /*--------------*/

  MPI_Finalize();
}

File: JICS_MPI_11/Fortran/group/group.f
      program group
      include 'mpif.h'

      integer NPROCS
      parameter(NPROCS=12)
      integer rank, new_rank, sendbuf, recvbuf, numtasks
      integer ranks1(6), ranks2(6), ierr
      integer orig_group, new_group, new_comm
      data ranks1 /0, 1, 2, 3, 4, 5/, ranks2 /6, 7, 8, 9, 10, 11/

C     *----------------*
C    * Initialize MPI *
C   *----------------*

      call MPI_INIT(ierr)

C     *------------------------------------------------------*
C    * Get my rank in the MPI_COMM_WORLD communicator group *
C   *------------------------------------------------------*

      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr)

C     *-------------------------------------------------------*
C    * Get the size of the MPI_COMM_WORLD communicator group *
C   *-------------------------------------------------------*

      call MPI_COMM_SIZE(MPI_COMM_WORLD, numtasks, ierr)

      if (numtasks .ne. NPROCS) then
        print *, 'Must specify MPROCS= ',NPROCS,' Terminating.'
        call MPI_FINALIZE(ierr)
        stop
      endif

      sendbuf = rank

C     *-----------------------------------*
C    * Extract the original group handle *
C   *-----------------------------------*

      call MPI_COMM_GROUP(MPI_COMM_WORLD, orig_group, ierr)

C     *-------------------------------------------------------*
C    * Divide tasks into two distinct groups based upon rank *
C   *-------------------------------------------------------*

      if  (rank .lt. NPROCS/2) then
         call MPI_GROUP_INCL(orig_group, NPROCS/2, ranks1, 
     &                  new_group, ierr)
      else 
      call MPI_GROUP_INCL(orig_group, NPROCS/2, ranks2, 
     &                  new_group, ierr)
      endif

C     *------------------------------------------------------------------------*
C    * Create new new communicator and then perform collective communications *
C   *------------------------------------------------------------------------*

      call MPI_COMM_CREATE(MPI_COMM_WORLD, new_group, 
     &                  new_comm, ierr)
      call MPI_ALLREDUCE(sendbuf, recvbuf, 1, MPI_INTEGER,
     &                  MPI_SUM, new_comm, ierr)

C     *-----------------------------------------------------*
C    * Get my rank in the new new_group communicator group *
C   *-----------------------------------------------------*

      call MPI_GROUP_RANK(new_group, new_rank, ierr)

      print *, 'rank= ',rank,' newrank= ',new_rank,' recvbuf= ',
     &     recvbuf

C     *--------------*
C    * Finalize MPI *
C   *--------------*

      call MPI_FINALIZE(ierr)

      end
